File: app/src/main/java/com/metahumanz/pacilread/reader/modern/dialog/ReaderDialogSupport.java
package com.metahumanz.pacilread.reader.modern.dialog;

import android.app.AlertDialog;
import android.content.res.Configuration;
import android.graphics.Color;
import android.graphics.Insets;
import android.graphics.Rect;
import android.graphics.drawable.ColorDrawable;
import android.os.Build;
import android.view.Gravity;
import android.view.View;
import android.view.ViewGroup;
import android.view.Window;
import android.view.WindowInsets;
import android.view.WindowInsetsController;
import android.view.WindowManager;
import android.widget.ArrayAdapter;
import android.widget.FrameLayout;
import android.widget.SeekBar;
import android.widget.TextView;

import androidx.core.view.WindowCompat;

import com.metahumanz.pacilread.R;
import com.metahumanz.pacilread.reader.modern.ModernReaderActivity;
import com.metahumanz.pacilread.reader.modern.ReaderRuntime;
import com.metahumanz.pacilread.reader.modern.ReaderUiUtils;
import com.metahumanz.pacilread.theme.ThemeModeHelper;
import com.metahumanz.pacilread.ui.GlassUiHelper;
import com.metahumanz.pacilread.ui.LaunchSourceTransition;
import com.metahumanz.pacilread.ui.PredictiveDialogDismissController;
import com.metahumanz.pacilread.ui.TransitionMotionModeHelper;

import java.util.List;

public final class ReaderDialogSupport {
    private final ModernReaderActivity activity;
    private final ReaderRuntime runtime;
    private final ReaderUiUtils ui;
    private LaunchSourceTransition.Source nextDismissSource;

    public ReaderDialogSupport(ModernReaderActivity activity, ReaderRuntime runtime, ReaderUiUtils ui) {
        this.activity = activity;
        this.runtime = runtime;
        this.ui = ui;
    }

    public ArrayAdapter<String> buildSpinnerAdapter(String[] items) {
        ArrayAdapter<String> adapter = new ArrayAdapter<>(activity, R.layout.item_spinner_selected, items);
        adapter.setDropDownViewResource(R.layout.item_spinner_dropdown);
        return adapter;
    }

    public ArrayAdapter<String> buildDialogListAdapter(List<String> items) {
        return new ArrayAdapter<String>(activity, R.layout.item_dialog_list_row, android.R.id.text1, items) {
            @Override
            public View getView(int position, View convertView, ViewGroup parent) {
                View view = super.getView(position, convertView, parent);
                TextView textView = view.findViewById(android.R.id.text1);
                textView.setTextColor(ui.themeColor(R.color.on_surface));
                return view;
            }
        };
    }

    public void setNextDismissSource(View sourceView) {
        nextDismissSource = LaunchSourceTransition.captureSource(sourceView);
    }

    public void setNextDismissSource(Rect sourceBounds) {
        nextDismissSource = sourceBounds == null ? null : LaunchSourceTransition.sourceFromBounds(sourceBounds);
    }

    public void setNextDismissSource(LaunchSourceTransition.Source source) {
        nextDismissSource = source;
    }

    public void showStyledDialog(AlertDialog dialog) {
        Window window = showDialogWithAnimation(dialog, R.style.ReaderPopDialogAnimation);
        if (window != null) {
            window.setBackgroundDrawableResource(android.R.color.transparent);
        }
        PredictiveDialogDismissController.Registration backRegistration = installPredictiveDismiss(dialog, window, consumeNextDismissSource());
        dialog.setOnDismissListener(unused -> backRegistration.unregister());
        GlassUiHelper.applyToHierarchy(activity, dialog.findViewById(android.R.id.content), runtime.settingsStore.getGlassOpacityPercent());
    }

    public void showFullscreenDialog(AlertDialog dialog) {
        Window window = showDialogWithAnimation(dialog, R.style.ReaderFullscreenDialogAnimation);
        if (window != null) {
            window.setBackgroundDrawableResource(android.R.color.transparent);
            window.setLayout(ViewGroup.LayoutParams.MATCH_PARENT, ViewGroup.LayoutParams.MATCH_PARENT);
        }
        PredictiveDialogDismissController.Registration backRegistration = installPredictiveDismiss(dialog, window, consumeNextDismissSource());
        dialog.setOnDismissListener(unused -> backRegistration.unregister());
        GlassUiHelper.applyToHierarchy(activity, dialog.findViewById(android.R.id.content), runtime.settingsStore.getGlassOpacityPercent());
    }

    public void applyTocStyleFullscreenInsets(View root, View contentContainer) {
        if (root == null || contentContainer == null) {
            return;
        }
        root.setOnApplyWindowInsetsListener((view, windowInsets) -> {
            int leftInset;
            int topInset;
            int rightInset;
            int bottomInset;
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.R) {
                Insets systemBars = windowInsets.getInsets(WindowInsets.Type.systemBars());
                Insets cutout = windowInsets.getInsets(WindowInsets.Type.displayCutout());
                boolean landscape = view.getResources().getConfiguration().orientation == Configuration.ORIENTATION_LANDSCAPE;
                leftInset = landscape ? systemBars.left : Math.max(systemBars.left, cutout.left);
                topInset = Math.max(systemBars.top, cutout.top);
                rightInset = landscape ? systemBars.right : Math.max(systemBars.right, cutout.right);
                bottomInset = Math.max(systemBars.bottom, cutout.bottom);
            } else {
                leftInset = windowInsets.getSystemWindowInsetLeft();
                topInset = windowInsets.getSystemWindowInsetTop();
                rightInset = windowInsets.getSystemWindowInsetRight();
                bottomInset = windowInsets.getSystemWindowInsetBottom();
            }
            contentContainer.setPadding(
                    ui.dp(20) + leftInset,
                    ui.dp(18) + topInset,
                    ui.dp(16) + rightInset,
                    ui.dp(16) + bottomInset
            );
            return windowInsets;
        });
    }

    public void addAlignedCloseButton(
            View contentView,
            int titleViewId,
            View contentContainer,
            AlertDialog dialog
    ) {
        if (!(contentView instanceof FrameLayout) || contentContainer == null || dialog == null) {
            return;
        }
        View titleView = contentView.findViewById(titleViewId);
        if (titleView == null) {
            return;
        }
        FrameLayout root = (FrameLayout) contentView;
        TextView closeButton = new TextView(activity);
        closeButton.setText("×");
        closeButton.setTextSize(20f);
        closeButton.setTextColor(ui.themeColor(R.color.on_surface));
        closeButton.setGravity(Gravity.CENTER);
        closeButton.setImportantForAccessibility(View.IMPORTANT_FOR_ACCESSIBILITY_YES);
        closeButton.setContentDescription("关闭");
        closeButton.setOnClickListener(v -> dialog.dismiss());

        int size = ui.dp(48);
        FrameLayout.LayoutParams params = new FrameLayout.LayoutParams(size, size);
        params.gravity = Gravity.TOP | Gravity.START;
        root.addView(closeButton, params);

        Runnable position = () -> positionAlignedCloseButton(root, titleView, contentContainer, closeButton, size);
        root.post(position);
        root.addOnLayoutChangeListener((view, left, top, right, bottom, oldLeft, oldTop, oldRight, oldBottom) -> position.run());
        titleView.addOnLayoutChangeListener((view, left, top, right, bottom, oldLeft, oldTop, oldRight, oldBottom) -> position.run());
        contentContainer.addOnLayoutChangeListener((view, left, top, right, bottom, oldLeft, oldTop, oldRight, oldBottom) -> position.run());
    }

    private void positionAlignedCloseButton(
            FrameLayout root,
            View titleView,
            View contentContainer,
            View closeButton,
            int size
    ) {
        if (root.getWidth() <= 0 || titleView.getWidth() <= 0 || contentContainer.getWidth() <= 0) {
            return;
        }
        int[] rootLocation = new int[2];
        int[] titleLocation = new int[2];
        int[] containerLocation = new int[2];
        root.getLocationOnScreen(rootLocation);
        titleView.getLocationOnScreen(titleLocation);
        contentContainer.getLocationOnScreen(containerLocation);

        int titleCenterY = titleLocation[1] - rootLocation[1] + titleView.getHeight() / 2;
        int contentRight = containerLocation[0] - rootLocation[0]
                + contentContainer.getWidth()
                - contentContainer.getPaddingRight();
        int left = Math.max(0, contentRight - size);
        int top = Math.max(0, titleCenterY - size / 2);

        FrameLayout.LayoutParams params = (FrameLayout.LayoutParams) closeButton.getLayoutParams();
        if (params.leftMargin != left || params.topMargin != top) {
            params.leftMargin = left;
            params.topMargin = top;
            closeButton.setLayoutParams(params);
        }
    }

    public void showImmersiveFullscreenDialog(AlertDialog dialog, boolean restoreShowSystemBars) {
        Window window = showDialogWithAnimation(dialog, R.style.ReaderFullscreenDialogAnimation);
        if (window != null) {
            window.setBackgroundDrawable(new ColorDrawable(Color.TRANSPARENT));
            window.getDecorView().setPadding(0, 0, 0, 0);
            window.setLayout(ViewGroup.LayoutParams.MATCH_PARENT, ViewGroup.LayoutParams.MATCH_PARENT);
            configureEdgeToEdgeWindow(window);
            applySystemBarsVisibility(window, false);
        }
        PredictiveDialogDismissController.Registration backRegistration = installPredictiveDismiss(dialog, window, consumeNextDismissSource());
        dialog.setOnDismissListener(unused -> {
            backRegistration.unregister();
            applySystemBarsVisibility(activity.getWindow(), restoreShowSystemBars);
        });
    }

    private Window showDialogWithAnimation(AlertDialog dialog, int animationStyleResId) {
        applyWindowAnimation(dialog, animationStyleResId);
        dialog.show();
        applyWindowAnimation(dialog, animationStyleResId);
        return dialog.getWindow();
    }

    private void applyWindowAnimation(AlertDialog dialog, int animationStyleResId) {
        Window window = dialog.getWindow();
        if (window != null) {
            window.setWindowAnimations(animationStyleResId);
        }
    }

    private LaunchSourceTransition.Source consumeNextDismissSource() {
        LaunchSourceTransition.Source source = nextDismissSource;
        nextDismissSource = null;
        return source;
    }

    private PredictiveDialogDismissController.Registration installPredictiveDismiss(
            AlertDialog dialog,
            Window window,
            LaunchSourceTransition.Source dismissSource
    ) {
        return PredictiveDialogDismissController.install(
                dialog,
                window,
                TransitionMotionModeHelper.isFluidMode(runtime.settingsStore),
                dismissSource
        );
    }

    private void configureEdgeToEdgeWindow(Window window) {
        WindowCompat.setDecorFitsSystemWindows(window, false);
        window.addFlags(WindowManager.LayoutParams.FLAG_DRAWS_SYSTEM_BAR_BACKGROUNDS);
        window.setStatusBarColor(Color.TRANSPARENT);
        window.setNavigationBarColor(Color.TRANSPARENT);
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.P) {
            WindowManager.LayoutParams attributes = window.getAttributes();
            attributes.layoutInDisplayCutoutMode = WindowManager.LayoutParams.LAYOUT_IN_DISPLAY_CUTOUT_MODE_SHORT_EDGES;
            window.setAttributes(attributes);
        }
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.Q) {
            window.setStatusBarContrastEnforced(false);
            window.setNavigationBarContrastEnforced(false);
        }
    }

    private void applySystemBarsVisibility(Window window, boolean showSystemBars) {
        View decorView = window.getDecorView();
        int flags = View.SYSTEM_UI_FLAG_LAYOUT_STABLE
                | View.SYSTEM_UI_FLAG_LAYOUT_FULLSCREEN
                | View.SYSTEM_UI_FLAG_LAYOUT_HIDE_NAVIGATION;
        if (!isDarkReaderUi()) {
            flags |= View.SYSTEM_UI_FLAG_LIGHT_STATUS_BAR;
            flags |= View.SYSTEM_UI_FLAG_LIGHT_NAVIGATION_BAR;
        }
        if (!showSystemBars) {
            flags |= View.SYSTEM_UI_FLAG_FULLSCREEN
                    | View.SYSTEM_UI_FLAG_HIDE_NAVIGATION
                    | View.SYSTEM_UI_FLAG_IMMERSIVE_STICKY;
        }
        decorView.setSystemUiVisibility(flags);
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.R) {
            WindowInsetsController controller = decorView.getWindowInsetsController();
            if (controller != null) {
                controller.setSystemBarsBehavior(WindowInsetsController.BEHAVIOR_SHOW_TRANSIENT_BARS_BY_SWIPE);
                if (showSystemBars) {
                    controller.show(WindowInsets.Type.statusBars() | WindowInsets.Type.navigationBars());
                } else {
                    controller.hide(WindowInsets.Type.statusBars() | WindowInsets.Type.navigationBars());
                }
            }
        }
    }

    private boolean isDarkReaderUi() {
        return ThemeModeHelper.MODE_DARK.equals(ThemeModeHelper.getResolvedReaderBucket(activity));
    }

    public static final class SimpleSeekListener implements SeekBar.OnSeekBarChangeListener {
        private final Runnable callback;

        public SimpleSeekListener(Runnable callback) {
            this.callback = callback;
        }

        @Override
        public void onProgressChanged(SeekBar seekBar, int progress, boolean fromUser) {
            callback.run();
        }

        @Override
        public void onStartTrackingTouch(SeekBar seekBar) {
        }

        @Override
        public void onStopTrackingTouch(SeekBar seekBar) {
        }
    }
}

File: app/src/main/kotlin/com/metahumanz/pacilread/reader/modern/dialog/ReaderDialogSupport.kt
package com.metahumanz.pacilread.reader.modern.dialog

import android.app.AlertDialog
import android.content.res.Configuration
import android.graphics.Color
import android.graphics.Rect
import android.graphics.drawable.ColorDrawable
import android.os.Build
import android.view.Gravity
import android.view.View
import android.view.ViewGroup
import android.view.Window
import android.view.WindowInsets
import android.view.WindowInsetsController
import android.view.WindowManager
import android.widget.ArrayAdapter
import android.widget.FrameLayout
import android.widget.SeekBar
import android.widget.TextView
import androidx.core.view.WindowCompat
import com.metahumanz.pacilread.R
import com.metahumanz.pacilread.reader.modern.ModernReaderActivity
import com.metahumanz.pacilread.reader.modern.ReaderRuntime
import com.metahumanz.pacilread.reader.modern.ReaderUiUtils
import com.metahumanz.pacilread.theme.ThemeModeHelper
import com.metahumanz.pacilread.ui.GlassUiHelper
import com.metahumanz.pacilread.ui.LaunchSourceTransition
import com.metahumanz.pacilread.ui.PredictiveDialogDismissController
import com.metahumanz.pacilread.ui.TransitionMotionModeHelper

class ReaderDialogSupport(
    private val activity: ModernReaderActivity,
    private val runtime: ReaderRuntime,
    private val ui: ReaderUiUtils,
) {
    private var nextDismissSource: LaunchSourceTransition.Source? = null

    fun buildSpinnerAdapter(items: Array<String>): ArrayAdapter<String> =
        ArrayAdapter(activity, R.layout.item_spinner_selected, items).apply {
            setDropDownViewResource(R.layout.item_spinner_dropdown)
        }

    fun buildDialogListAdapter(items: List<String>): ArrayAdapter<String> =
        object : ArrayAdapter<String>(activity, R.layout.item_dialog_list_row, android.R.id.text1, items) {
            override fun getView(position: Int, convertView: View?, parent: ViewGroup): View =
                super.getView(position, convertView, parent).also { view ->
                    view.findViewById<TextView>(android.R.id.text1).setTextColor(ui.themeColor(R.color.on_surface))
                }
        }

    fun setNextDismissSource(sourceView: View?) {
        nextDismissSource = LaunchSourceTransition.captureSource(sourceView)
    }

    fun setNextDismissSource(sourceBounds: Rect?) {
        nextDismissSource = sourceBounds?.let(LaunchSourceTransition::sourceFromBounds)
    }

    fun setNextDismissSource(source: LaunchSourceTransition.Source?) {
        nextDismissSource = source
    }

    fun showStyledDialog(dialog: AlertDialog) {
        val window = showDialogWithAnimation(dialog, R.style.ReaderPopDialogAnimation)
        window?.setBackgroundDrawableResource(android.R.color.transparent)
        val registration = installPredictiveDismiss(dialog, window, consumeNextDismissSource())
        dialog.setOnDismissListener { registration.unregister() }
        GlassUiHelper.applyToHierarchy(activity, dialog.findViewById(android.R.id.content), runtime.settingsStore.glassOpacityPercent)
    }

    fun showFullscreenDialog(dialog: AlertDialog) {
        val window = showDialogWithAnimation(dialog, R.style.ReaderFullscreenDialogAnimation)
        window?.apply {
            setBackgroundDrawableResource(android.R.color.transparent)
            setLayout(ViewGroup.LayoutParams.MATCH_PARENT, ViewGroup.LayoutParams.MATCH_PARENT)
        }
        val registration = installPredictiveDismiss(dialog, window, consumeNextDismissSource())
        dialog.setOnDismissListener { registration.unregister() }
        GlassUiHelper.applyToHierarchy(activity, dialog.findViewById(android.R.id.content), runtime.settingsStore.glassOpacityPercent)
    }

    @Suppress("DEPRECATION")
    fun applyTocStyleFullscreenInsets(root: View?, contentContainer: View?) {
        if (root == null || contentContainer == null) return
        root.setOnApplyWindowInsetsListener { view, windowInsets ->
            val leftInset: Int
            val topInset: Int
            val rightInset: Int
            val bottomInset: Int
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.R) {
                val systemBars = windowInsets.getInsets(WindowInsets.Type.systemBars())
                val cutout = windowInsets.getInsets(WindowInsets.Type.displayCutout())
                val landscape = view.resources.configuration.orientation == Configuration.ORIENTATION_LANDSCAPE
                leftInset = if (landscape) systemBars.left else Math.max(systemBars.left, cutout.left)
                topInset = Math.max(systemBars.top, cutout.top)
                rightInset = if (landscape) systemBars.right else Math.max(systemBars.right, cutout.right)
                bottomInset = Math.max(systemBars.bottom, cutout.bottom)
            } else {
                leftInset = windowInsets.systemWindowInsetLeft
                topInset = windowInsets.systemWindowInsetTop
                rightInset = windowInsets.systemWindowInsetRight
                bottomInset = windowInsets.systemWindowInsetBottom
            }
            contentContainer.setPadding(ui.dp(20) + leftInset, ui.dp(18) + topInset, ui.dp(16) + rightInset, ui.dp(16) + bottomInset)
            windowInsets
        }
    }

    fun addAlignedCloseButton(contentView: View?, titleViewId: Int, contentContainer: View?, dialog: AlertDialog?) {
        if (contentView !is FrameLayout || contentContainer == null || dialog == null) return
        val titleView = contentView.findViewById<View>(titleViewId) ?: return
        val closeButton = TextView(activity).apply {
            text = "×"
            textSize = 20f
            setTextColor(ui.themeColor(R.color.on_surface))
            gravity = Gravity.CENTER
            importantForAccessibility = View.IMPORTANT_FOR_ACCESSIBILITY_YES
            contentDescription = "关闭"
            setOnClickListener { dialog.dismiss() }
        }
        val size = ui.dp(48)
        contentView.addView(closeButton, FrameLayout.LayoutParams(size, size).apply { gravity = Gravity.TOP or Gravity.START })
        val position = Runnable { positionAlignedCloseButton(contentView, titleView, contentContainer, closeButton, size) }
        contentView.post(position)
        val listener = View.OnLayoutChangeListener { _, _, _, _, _, _, _, _, _ -> position.run() }
        contentView.addOnLayoutChangeListener(listener)
        titleView.addOnLayoutChangeListener(listener)
        contentContainer.addOnLayoutChangeListener(listener)
    }

    private fun positionAlignedCloseButton(root: FrameLayout, titleView: View, contentContainer: View, closeButton: View, size: Int) {
        if (root.width <= 0 || titleView.width <= 0 || contentContainer.width <= 0) return
        val rootLocation = IntArray(2)
        val titleLocation = IntArray(2)
        val containerLocation = IntArray(2)
        root.getLocationOnScreen(rootLocation)
        titleView.getLocationOnScreen(titleLocation)
        contentContainer.getLocationOnScreen(containerLocation)
        val titleCenterY = titleLocation[1] - rootLocation[1] + titleView.height / 2
        val contentRight = containerLocation[0] - rootLocation[0] + contentContainer.width - contentContainer.paddingRight
        val left = Math.max(0, contentRight - size)
        val top = Math.max(0, titleCenterY - size / 2)
        val params = closeButton.layoutParams as FrameLayout.LayoutParams
        if (params.leftMargin != left || params.topMargin != top) {
            params.leftMargin = left
            params.topMargin = top
            closeButton.layoutParams = params
        }
    }

    fun showImmersiveFullscreenDialog(dialog: AlertDialog, restoreShowSystemBars: Boolean) {
        val window = showDialogWithAnimation(dialog, R.style.ReaderFullscreenDialogAnimation)
        window?.apply {
            setBackgroundDrawable(ColorDrawable(Color.TRANSPARENT))
            decorView.setPadding(0, 0, 0, 0)
            setLayout(ViewGroup.LayoutParams.MATCH_PARENT, ViewGroup.LayoutParams.MATCH_PARENT)
            configureEdgeToEdgeWindow(this)
            applySystemBarsVisibility(this, false)
        }
        val registration = installPredictiveDismiss(dialog, window, consumeNextDismissSource())
        dialog.setOnDismissListener {
            registration.unregister()
            applySystemBarsVisibility(activity.window, restoreShowSystemBars)
        }
    }

    private fun showDialogWithAnimation(dialog: AlertDialog, animationStyleResId: Int): Window? {
        applyWindowAnimation(dialog, animationStyleResId)
        dialog.show()
        applyWindowAnimation(dialog, animationStyleResId)
        return dialog.window
    }

    private fun applyWindowAnimation(dialog: AlertDialog, animationStyleResId: Int) {
        dialog.window?.setWindowAnimations(animationStyleResId)
    }

    private fun consumeNextDismissSource(): LaunchSourceTransition.Source? = nextDismissSource.also { nextDismissSource = null }

    private fun installPredictiveDismiss(
        dialog: AlertDialog, window: Window?, dismissSource: LaunchSourceTransition.Source?,
    ): PredictiveDialogDismissController.Registration = PredictiveDialogDismissController.install(
        dialog, window, TransitionMotionModeHelper.isFluidMode(runtime.settingsStore), dismissSource,
    )

    @Suppress("DEPRECATION")
    private fun configureEdgeToEdgeWindow(window: Window) {
        WindowCompat.setDecorFitsSystemWindows(window, false)
        window.addFlags(WindowManager.LayoutParams.FLAG_DRAWS_SYSTEM_BAR_BACKGROUNDS)
        window.statusBarColor = Color.TRANSPARENT
        window.navigationBarColor = Color.TRANSPARENT
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.P) {
            window.attributes = window.attributes.apply {
                layoutInDisplayCutoutMode = WindowManager.LayoutParams.LAYOUT_IN_DISPLAY_CUTOUT_MODE_SHORT_EDGES
            }
        }
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.Q) {
            window.isStatusBarContrastEnforced = false
            window.isNavigationBarContrastEnforced = false
        }
    }

    @Suppress("DEPRECATION")
    private fun applySystemBarsVisibility(window: Window, showSystemBars: Boolean) {
        val decorView = window.decorView
        var flags = View.SYSTEM_UI_FLAG_LAYOUT_STABLE or View.SYSTEM_UI_FLAG_LAYOUT_FULLSCREEN or View.SYSTEM_UI_FLAG_LAYOUT_HIDE_NAVIGATION
        if (!isDarkReaderUi()) flags = flags or View.SYSTEM_UI_FLAG_LIGHT_STATUS_BAR or View.SYSTEM_UI_FLAG_LIGHT_NAVIGATION_BAR
        if (!showSystemBars) flags = flags or View.SYSTEM_UI_FLAG_FULLSCREEN or View.SYSTEM_UI_FLAG_HIDE_NAVIGATION or View.SYSTEM_UI_FLAG_IMMERSIVE_STICKY
        decorView.systemUiVisibility = flags
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.R) {
            decorView.windowInsetsController?.apply {
                systemBarsBehavior = WindowInsetsController.BEHAVIOR_SHOW_TRANSIENT_BARS_BY_SWIPE
                val types = WindowInsets.Type.statusBars() or WindowInsets.Type.navigationBars()
                if (showSystemBars) show(types) else hide(types)
            }
        }
    }

    private fun isDarkReaderUi(): Boolean = ThemeModeHelper.MODE_DARK == ThemeModeHelper.getResolvedReaderBucket(activity)

    class SimpleSeekListener(private val callback: Runnable) : SeekBar.OnSeekBarChangeListener {
        override fun onProgressChanged(seekBar: SeekBar?, progress: Int, fromUser: Boolean) = callback.run()
        override fun onStartTrackingTouch(seekBar: SeekBar?) = Unit
        override fun onStopTrackingTouch(seekBar: SeekBar?) = Unit
    }
}
